com.sap.cloud.lm.sl.cf.process/src/main/java/com/sap/cloud/lm/sl/cf/process/flowable/FlowableFacade.java

```diff
@@ -226,42 +241,128 @@ public void trigger(String executionId, Map<String, Object> variables) {
     }
 
     public void deleteProcessInstance(String processInstanceId, String deleteReason) {
-        long overallAbortDeadline = System.currentTimeMillis() + DEFAULT_ABORT_WAIT_TIMEOUT_MS + DEFAULT_ABORT_TIMEOUT_MS;
+        long overallAbortDeadline = System.currentTimeMillis() + DEFAULT_ABORT_TIMEOUT_MS;
         while (true) {
             try {
                 if (isProcessInstanceAtReceiveTask(processInstanceId)) {
-                    processEngine.getRuntimeService()
-                                 .deleteProcessInstance(processInstanceId, deleteReason);
-                    break;
-                }
-                
-                LOGGER.debug(format(Messages.SETTING_VARIABLE, Constants.PROCESS_ABORTED, Boolean.TRUE));
-
-                // TODO: Use execution ID instead of process instance ID, as
-                // they can be
-                // different if the process has parallel executions.
-                processEngine.getRuntimeService()
-                             .setVariable(processInstanceId, Constants.PROCESS_ABORTED, Boolean.TRUE);
-
-                long allSubprocessesFinishedDeadline = System.currentTimeMillis() + DEFAULT_ABORT_WAIT_TIMEOUT_MS;
-                while (true) {
-                    List<Execution> subprocessExecutionsWithoutChildren = getAllSubprocessExecutionsWithoutChildren(processInstanceId);
-
-                    if (allExecutionsHaveDeadLetterJobs(subprocessExecutionsWithoutChildren)
-                        || isPastDeadline(allSubprocessesFinishedDeadline)) {
-                        break;
-                    }
+                    LOGGER.info(format(PROCESS_INSTANCE_0_IS_AT_RECEIVE_TASK, processInstanceId));
+                    abortProcessInstance(processInstanceId, deleteReason);
+                    return;
                 }
-                processEngine.getRuntimeService()
-                             .deleteProcessInstance(processInstanceId, deleteReason);
-                break;
+                abortProcessSafely(processInstanceId, deleteReason);
+                return;
             } catch (FlowableOptimisticLockingException e) {
-                if (isPastDeadline(overallAbortDeadline)) {
-                    throw new IllegalStateException(Messages.ABORT_OPERATION_TIMED_OUT, e);
-                }
-                LOGGER.warn(format(Messages.RETRYING_PROCESS_ABORT, processInstanceId));
+                throwExceptionInCaseOfTimeout(processInstanceId, overallAbortDeadline, e);
+            } catch (NotFoundException e) {
+                LOGGER.error(e.getMessage(), e);
+                return;
+            }
+        }
+    }
+
+    private void abortProcessSafely(String processInstanceId, String deleteReason) {
```

Comment:
!!!!!Check this comment last!!!!!

Overall, the idea is the same which was firstly proposed. However, I do not like the approach here because you are making while cycle inside another while cycle -> this is a lot of complexity. What I would suggest is to to make the while cycle with condition: while(!executions.isEmpty()) ... and put in this executions list whenever you find new execution that need to be checked. If you want we could discuss this further..

Reply:
Done